### ISSUE TITLE
Close the CUDA reduction, statistics, and order-statistic gap (14 operations)

### ISSUE THREAD

**Issue by xupremix** [xupremix/incin#87]
## Problem

CUDA has `pointwise` and `reduction` profiles advertised, but the reduction coverage stops at sum/mean/max/min. Fourteen catalog reductions and order statistics have no CUDA executor:

- **Statistics (9):** `var_all`, `var_dim`, `var_keepdim`, `std_all`, `std_dim`, `std_keepdim`, `norm`, `prod_all`, `prod_dim`
- **Order and index (5):** `argmax`, `argmin`, `argsort`, `topk`, `cumsum`

These are not interchangeable with the sum reductions. `argmax` is how a classifier reports a prediction, `topk` is how a mixture-of-experts router selects experts, `cumsum` is how a router computes expert offsets, and `var`/`std` are what normalization and metrics are made of. Their absence is why a CUDA model can compute a forward pass and still not produce an answer without a device-to-host round trip.

`topk` is additionally CPU-only and `f32`-only even there, which makes it the narrowest operation in the set relative to how central it is to the routing work.

## Proposed implementation

- Implement the statistics reductions with a numerically stable single-pass formulation, accumulating in f32 for lower-precision inputs, and share one reduction skeleton across all-axis, per-axis, and keepdim forms rather than writing three kernels per operation.
- Implement `cumsum` as a work-efficient scan rather than a serial pass, since expert-offset computation runs it on the batch dimension.
- Implement `argmax`, `argmin`, and `topk` with index-carrying reductions, and define the tie-breaking rule explicitly in `docs/OPERATION_SEMANTICS.md` so CPU and CUDA agree rather than agreeing by accident.
- Implement `argsort` over the segment sizes routing actually uses, and record the stability guarantee.
- Widen `topk` past `f32` on both CPU and CUDA while implementing it here.

## Acceptance criteria

- All 14 operations execute on CUDA and match the CPU reference, including index outputs, not only values.
- Tie-breaking for `argmax`, `argmin`, `topk`, and `argsort` is documented and identical between CPU and CUDA.
- `var` and `std` match CPU within documented tolerances for inputs where a naive two-pass formulation would lose precision.
- Gradients match the CPU reference for every row advertised as training-capable.
- `topk` advertises the same dtype set as the other order statistics on both backends.

## Proposed API and capability-row change

No new public API for 13 of the 14. The exception is `topk`, whose dtype set widens on CPU as well.

```rust
use incin::prelude::*;
type Backend = DefaultBackend;   // with `cuda`, the CUDA backend

let logits: Tensor<s![32, 1000], Backend, f32> = model.forward(x)?;
let pred = logits.argmax(1)?;                 // today on CUDA: capability refusal
let (vals, idx) = logits.topk(5, 1, true)?;   // today: CPU only, f32 only

// A CUDA classifier currently cannot report its own prediction without a
// device-to-host round trip, which is the practical cost of this gap.
```

`topk` is the narrowest operation in the catalog relative to how central it is: CPU-only and `f32`-only, while being the core primitive of mixture-of-experts routing.

## Shape and type-system invariants

| Operation | Shape rule | Contract that must be pinned down |
|---|---|---|
| `var_all`, `std_all`, `prod_all` | `[..] -> []` | Accumulate in f32 for reduced-precision inputs |
| `var_dim`, `std_dim`, `prod_dim` | reduce `dim` | Follows the existing `*_dim` convention |
| `var_keepdim`, `std_keepdim` | reduce `dim`, keep rank | |
| `norm` | reduce, `p` as an attribute | |
| `argmax`, `argmin` | `[..] -> [..]` minus `dim`, dtype `i64` | **Tie-breaking** |
| `topk(k, dim)` | `[.., D, ..] -> [.., k, ..]`, returns `(values, indices)` | **Tie-breaking**; `k <= D` |
| `argsort(dim, desc)` | shape-preserving, dtype `i64` | **Stability** |
| `cumsum(dim)` | shape-preserving | Accumulation order affects the last elements |

The `IndexResult` dtype rule already exists in `crates/incin-core/src/exec/catalog/classification.rs`, so the index-returning operations have their output dtype fixed by the catalog rather than by each backend.

**Tie-breaking is the one contract that must be written down before implementation, not after.** CPU and CUDA agreeing by accident is not the same as agreeing, and a conformance failure caused by an unspecified tie is indistinguishable from a real bug.

## Background

- **Statistics need a stable single-pass form.** A naive two-pass mean-then-variance is fine on CPU where memory is cheap and catastrophic in a GPU reduction that wants one pass. Welford or a shifted-data formulation, with f32 accumulation regardless of input dtype.
- **`cumsum` must be a work-efficient scan, not a serial pass.** Expert-offset computation (#103) runs it on the batch dimension, so a serial implementation is a sequential bottleneck in the middle of a parallel routing step.
- **Index-carrying reductions are a different kernel shape.** `argmax` reduces a `(value, index)` pair with a comparison that must break ties consistently across the tree, which is where the tie-breaking contract stops being academic: a tree reduction's association order is not the array order.
- **`argsort` at routing segment sizes.** The relevant regime is sorting a few thousand keys per step, repeatedly, not a one-off sort of a huge array. That favors a segmented approach over a general-purpose sort.

## How this interoperates

- **#103** needs `topk`, `cumsum`, and `argsort` on device for mixture-of-experts routing. This issue and #88 are the CUDA half of the routing surface; #103 is the missing-operations half.
- **#83** verifies index outputs, not only values, which is the part a hand-written test usually skips.
- **`docs/OPERATION_SEMANTICS.md`** is where the tie-breaking and stability guarantees land, and it is the document `#83` compares against.
- **#90** widens `topk` and the statistics past `f32`.
- **#84** needs stable statistics for `group_norm` and `instance_norm`, so the reduction skeleton is shared work.
- **`docs/capabilities.md`** regenerates from the registration.

## Alternatives

Falling back to a host round trip for the index reductions would make them nominally available sooner, but a per-step device-to-host copy in a training loop is a performance defect rather than a capability.

## Additional context

Relevant code: `crates/incin-backends/src/cuda/ops/reduce.rs`, `crates/incin-backends/src/cuda/ops/kernels/`, `docs/OPERATION_SEMANTICS.md`, and the CPU references under `crates/incin-backends/src/cpu/ops/`.

`topk`, `cumsum`, and `argsort` are prerequisites for the mixture-of-experts routing work in this milestone.

**Comment by xupremix** [comment]
Rechecked against `master`. **All 14 operations named here are now advertised on CUDA** — all nine statistics and all five order/index rows, verified individually.

One acceptance criterion is objectively checkable without #83, and it is **not met** — in the opposite direction to the one described:

| | CPU | CUDA |
|---|---|---|
| `topk` | `f32` | `bf16, f16, f32, f64` |
| `argmax` / `argmin` / `argsort` | 8 dtypes | `f32` |

The criterion asks that `topk` advertise the same dtype set as the other order statistics on both backends. It matches on neither. `topk` is still `f32`-only on CPU as the problem statement says, but on CUDA it is now *wider* than `argmax`/`argmin`/`argsort` rather than narrower. Whatever is settled here should account for both directions.

The remaining criteria — value and index agreement with CPU, documented tie-breaking, `var`/`std` precision, gradient agreement — all need execution evidence and so wait on #83.
